### PR TITLE
[mlir] [dataflow] Refactoring the definition of program points in dat…

### PR DESCRIPTION
**Pull request by cxy-1993**
…a flow analysis

This patch distinguishes between program points and lattice anchors in data flow analysis, where lattice anchors represent locations where a lattice can be attached, while program points denote points in program execution.

Related discussions: https://discourse.llvm.org/t/rfc-unify-the-semantics-of-program-points/80671/8

### CODE REVIEW

**Review by Mogball — APPROVED** [review]
LGTM, thanks for driving this!

**Comment by MaheshRavishankar** [comment]
This change seems to have broken `torch-mlir`. I fully expect that this is probably an issue in torch-mlir itself, but The way the data flow framework is used in [`InlineGlobalSlots.cpp`](https://github.com/llvm/torch-mlir/blob/main/lib/Dialect/Torch/Transforms/InlineGlobalSlots.cpp) seems incompatible with this change. I havent been able to make changes to this to get this to build. 

I dont have a repro or anything, and I am not fully upto speed on the DataFlow analysis or how it is used in Torch-MLIR but just leaving a note here that this seems to cause a breakage (due to potentially mis-use of the data flow framework)

**Comment by MaheshRavishankar** [comment]
Cross listing this change with IREE PR that brings in this MLIR change (https://github.com/iree-org/iree/pull/18384) for reference

**Comment by cxy-1993** [comment]
> This change seems to have broken `torch-mlir`. I fully expect that this is probably an issue in torch-mlir itself, but The way the data flow framework is used in [`InlineGlobalSlots.cpp`](https://github.com/llvm/torch-mlir/blob/main/lib/Dialect/Torch/Transforms/InlineGlobalSlots.cpp) seems incompatible with this change. I havent been able to make changes to this to get this to build.
> 
> I dont have a repro or anything, and I am not fully upto speed on the DataFlow analysis or how it is used in Torch-MLIR but just leaving a note here that this seems to cause a breakage (due to potentially mis-use of the data flow framework)

Thanks for the feedback. It's clear that our current analysis doesn't align well with the patch's modifications. After a brief review of the code (please point out any errors), I believe this analysis is essentially a more robust version of sparse data flow analysis, extended to track attrs. The distinguishing characteristic is that the visit process might be triggered by an attr (symbol ref) rather than an operation.

In my opinion, this doesn't fundamentally deviate from the essence of current data flow analysis. When we update flat symbol attrs or op result, we still rely on initializeGlobalSlotsOp/GlobalSlotGetOp to update operand states. It's just that the current updates only modify the states of some operands in initializeGlobalSlotsOp(visit op will update all). Therefore, I suggest modifying the existing visit op result at (https://github.com/llvm/torch-mlir/blob/98e08023bbf71e00ab81e980eac9f7c96f1f24b4/lib/Dialect/Torch/Transforms/InlineGlobalSlots.cpp#L193C31-L193C46) to visit GlobalSlotGetOp, and modifying the visit flatSymbolRefPoint at (https://github.com/llvm/torch-mlir/blob/98e08023bbf71e00ab81e980eac9f7c96f1f24b4/lib/Dialect/Torch/Transforms/InlineGlobalSlots.cpp#L210) to visit initializeGlobalSlotsOp. This would align with the current dataflow analysis's definition of a program point.

If you have any further questions, please don't hesitate to discuss them with me. Additionally, I'm more than happy to assist you in modifying this section of code if needed.

I believe this modification is a step in the right direction for data flow analysis, and I hope it can be accepted by everyone. I plan to further standardize the semantics of program points, which may introduce some incompatibilities. I'm more than happy to help fix any issues that may arise.

**Comment by marbre** [comment]
> In my opinion, this doesn't fundamentally deviate from the essence of current data flow analysis. When we update flat symbol attrs or op result, we still rely on initializeGlobalSlotsOp/GlobalSlotGetOp to update operand states. It's just that the current updates only modify the states of some operands in initializeGlobalSlotsOp(visit op will update all). Therefore, I suggest modifying the existing visit op result at (https://github.com/llvm/torch-mlir/blob/98e08023bbf71e00ab81e980eac9f7c96f1f24b4/lib/Dialect/Torch/Transforms/InlineGlobalSlots.cpp#L193C31-L193C46) to visit GlobalSlotGetOp, and modifying the visit flatSymbolRefPoint at (https://github.com/llvm/torch-mlir/blob/98e08023bbf71e00ab81e980eac9f7c96f1f24b4/lib/Dialect/Torch/Transforms/InlineGlobalSlots.cpp#L210) to visit initializeGlobalSlotsOp. This would align with the current dataflow analysis's definition of a program point.
>
> If you have any further questions, please don't hesitate to discuss them with me. Additionally, I'm more than happy to assist you in modifying this section of code if needed.

Thanks for your suggestion! I am currently looking into how to adapt torch-mlir but I am unfortunately not familiar with the data-flow analysis and might have missed something. Therefore please allow to ask for further clarification.

While the argument to the `visit(ProgramPoint point)` method still is of a `ProgramPoint`, the latter was changed and now essentially is a `PointerUnion`. Before this change, `point` could be casted to a `Value` (within `visit()`) which is no longer possible. To get the value, one could instead cast `point` to an `Operation` and obtain the result value from this op. Furthermore, `getProgramPoint` needs to be replaced with `getLatticeAnchor`, thus
```C++
auto *flatSymbolRefPoint = getProgramPoint<FlatSymbolRefProgramPoint>(globalSlotGet.getSlotAttr());
auto *valueState = getOrCreateFor<InlineGlobalSlotsAnalysisState>(flatSymbolRefPoint, globalSlotGet.getResult()):
```
becomes
```C++
auto *flatSymbolRefPoint = getLatticeAnchor<FlatSymbolRefProgramPoint>(globalSlotGet.getSlotAttr());
auto *valueState = getOrCreateFor<InlineGlobalSlotsAnalysisState>(*flatSymbolRefPoint, globalSlotGet.getResult()):
```
What is failing here is `getOrCreateFor` as the first argument needs to be a `ProgramPoint` but now is a `FlatSymbolRefProgramPoint`. Any pointers would be appreciated!

**Comment by cxy-1993** [comment]
> > In my opinion, this doesn't fundamentally deviate from the essence of current data flow analysis. When we update flat symbol attrs or op result, we still rely on initializeGlobalSlotsOp/GlobalSlotGetOp to update operand states. It's just that the current updates only modify the states of some operands in initializeGlobalSlotsOp(visit op will update all). Therefore, I suggest modifying the existing visit op result at (https://github.com/llvm/torch-mlir/blob/98e08023bbf71e00ab81e980eac9f7c96f1f24b4/lib/Dialect/Torch/Transforms/InlineGlobalSlots.cpp#L193C31-L193C46) to visit GlobalSlotGetOp, and modifying the visit flatSymbolRefPoint at (https://github.com/llvm/torch-mlir/blob/98e08023bbf71e00ab81e980eac9f7c96f1f24b4/lib/Dialect/Torch/Transforms/InlineGlobalSlots.cpp#L210) to visit initializeGlobalSlotsOp. This would align with the current dataflow analysis's definition of a program point.
> > If you have any further questions, please don't hesitate to discuss them with me. Additionally, I'm more than happy to assist you in modifying this section of code if needed.
> 
> Thanks for your suggestion @cxy-1993! I am currently looking into how to adapt torch-mlir but I am unfortunately not familiar with the data-flow analysis and might have missed something. Therefore please allow to ask for further clarification.
> 
> While the argument to the `visit(ProgramPoint point)` method still is of a `ProgramPoint`, the latter was changed and now essentially is a `PointerUnion`. Before this change, `point` could be casted to a `Value` (within `visit()`) which is no longer possible. To get the value, one could instead cast `point` to an `Operation` and obtain the result value from this op. Furthermore, `getProgramPoint` needs to be replaced with `getLatticeAnchor`, thus
> 
> ```c++
> auto *flatSymbolRefPoint = getProgramPoint<FlatSymbolRefProgramPoint>(globalSlotGet.getSlotAttr());
> auto *valueState = getOrCreateFor<InlineGlobalSlotsAnalysisState>(flatSymbolRefPoint, globalSlotGet.getResult()):
> ```
> 
> becomes
> 
> ```c++
> auto *flatSymbolRefPoint = getLatticeAnchor<FlatSymbolRefProgramPoint>(globalSlotGet.getSlotAttr());
> auto *valueState = getOrCreateFor<InlineGlobalSlotsAnalysisState>(*flatSymbolRefPoint, globalSlotGet.getResult()):
> ```
> 
> What is failing here is `getOrCreateFor` as the first argument needs to be a `ProgramPoint` but now is a `FlatSymbolRefProgramPoint`, while assuming that the constructor of `InlineGlobalSlotsAnalysisState` needs to be changed from
> 
> ```c++
> InlineGlobalSlotsAnalysisState(ProgramPoint point) : AnalysisState(point) {
> ```
> 
> to
> 
> ```c++
> InlineGlobalSlotsAnalysisState(LatticeAnchor anchor) : AnalysisState(anchor) {
> ```
> 
> Any pointers would be appreciated!

I have a limited understanding of the specific functionality of this pass. Please point out any issues if you find any. For the sake of clarity, let's call the program point before this patch the 'ori program point'. My overall modification idea is as follows:

The InlineGlobalSlotsAnalysis::visit function might accept two types of ori program points and update the lattice state: one is the result of Torch::GlobalSlotGetOp, and the other is FlatSymbolRefProgramPoint. They will respectively use Torch::GlobalSlotGetOp and Torch::InitializeGlobalSlotsOp to update the states of other lattices.

Therefore, I suggest modifying the original code:

getOrCreateFor<InlineGlobalSlotsAnalysisState>(flatSymbolRefPoint, xxx) to getOrCreateFor<InlineGlobalSlotsAnalysisState>(initializeGlobalSlotsOp, xxx).
getOrCreateFor<InlineGlobalSlotsAnalysisState>(value, xx) to something like getOrCreateFor<InlineGlobalSlotsAnalysisState>(value.getDefiningOp<Torch::GlobalSlotGetOp>(), flatSymbolRefPoint).

At this point, the visit's program point will only have to deal initializeGlobalSlotsOp and GlobalSlotGetOp, and we can use the information from these two ops to update the lattice state related to these op.

**Comment by rsuderman** [comment]
So I believe we will want to revert this and reland in the future with changes.

The main issue is the changes to `ProgramPoint`. Where previously we allowed it to be an old ProgramPoint (now LatticeAnchor) it now assumes that the location is an operation or block. However the `torch-mlir` pass uses symnames instead to check for cases. Other frameworks do similar cases of `symnames` for referencing a shared handle not mapped to any particular operation. Likely the right choice is to remove `ProgramPoint` and have the dependent location be some type of LatticeAnchor or generalized type rather than hardcoding to an `Operation*` or `Block*`. If we don't want to assume the same type as `LatticeAnchor` we can create a similar generic for storing the program point used to visit.

In short, this actually removes some functionality that would require completely reworking downstream analysis passes to correct without there being a clear benefit in imposing this new restriction.

**Comment by stellaraccident** [comment]
Flyby comment 1: I am almost certain that all of the features behind InlineGlobalSlots.cpp are very close [to being decommissioned](https://github.com/llvm/torch-mlir/pull/3693), and then this very old code can be outright deleted vs carried forward. We've been clear since January that the day will come soon when the cost of maintaining this path cannot be carried anymore. If this is that day, then delete any remaining tests and I will approve.

Flyby comment 2: While we may get lucky in this case and simply not need to address this, in general, when doing an in-situ upgrade of an API, it is important to not just address where things get to but to not remove semantics that people are relying on along the way. I know this is hard/annoying. For this very specific case of impact to torch-mlir, we may be able to not worry about this at this time, but torch-mlir is not the only user of these APIs, and we need to be evolving them without semantic/representability reduction in intermediate steps. If that isn't possible, then it strongly indicates that an in-situ refactor isn't the appropriate way to go.

**Comment by Mogball** [comment]
This was the RFC https://discourse.llvm.org/t/rfc-unify-the-semantics-of-program-points/80671. Not many people engaged on the RFC. Waiting to be broken downstream before entering the discussion is probably not the ideal situation, but maybe we should have been more aggressive in pinging potential stakeholders.

That being said, I am sympathetic to the loss of functionality here (my approval was based on the supposed acceptance of the RFC and the code changes themselves). I am +1 to reverting this change but I think we should have a follow-up discussion on ProgramPoint afterwards. Maybe we can have an ODM for the first time in ages!

**Comment by stellaraccident** [comment]
From my perspective, this falls into "these things happen" bucket. I read the RFC a few times and failed to see the issue. I wasn't calling for any action this time but more just commenting for the future. I had forgotten that this use existed in this very old code path and didn't think to check deeply. It can be hard to fight inertia on this stuff, and I appreciate the effort it takes. 

Maybe give it another day? I still think we can just outright delete the live use downstream. I can't speak for whether other used this feature of the API. So it's a judgment call as to whether to proceed as is or better understand this detail and address with a revert.

**Comment by stellaraccident** [comment]
No apology necessary -- we've all done it. Especially for these older, established APIs, it is a matter of experience to know how to sequence changes in a way that works well for everyone. Generally, we are fine pushing "syntax" changes downstream but removing functionality without replacement is harder to deal with. It's tricky because sometimes it's not obvious that some corner case for used in some way. 

I was mainly being explicit about my comments so that we could learn for next time vs suggesting that we need to backtrack this time. The contributions are appreciated.

**Comment by joker-eph** [comment]
> We definitely need a way to notify all downstream stakeholders about the necessary infrastructure changes. I have another related commit in development, and I'm curious about what measures we can take to ensure it doesn't break compatibility with downstream repos (for example, perhaps creating a new dataflow directory for this commit?). And I'm also wondering at what point it would be safe to break compatibility.Any help is appreciated.

You did nothing wrong, it's fine. We break APIs all the time.
It's also hard to evolve components without breaking any use-cases, usually downstream find a way to update but sometimes we miss one use-case that can't be implemented anymore. In such case we just revert and work it out collaboratively (downstream are expected to actively contribute to the solution) to be able to re-land.

Downstream should implement better upstream test to model their use-cases if they want these to be better supported/maintained, otherwise they deal with occasional breakages.
